core/src/read.jl

```diff
@@ -1227,46 +1238,146 @@ function FlowDemand(db::DB, config::Config)::FlowDemand
 
-    subgrid_ids = Int32[]
-    basin_index = Int32[]
-    interpolations = ScalarInterpolation[]
-    has_error = false
-    for group in IterTools.groupby(row -> row.subgrid_id, tables)
+    # Since not all Basins need to have subgrids, don't enforce completeness.
+    _, _, _, valid =
+        static_and_time_node_ids(db, static, time, "Basin"; is_complete = false)
+    if !valid
+        error("Problems encountered when parsing Subgrid static and time node IDs.")
+    end
+
+    node_to_basin = Dict{Int32, Int}(
+        Int32(node_id) => index for (index, node_id) in enumerate(basin.node_id)
+    )
+    subgrid_id_static = Int32[]
+    basin_index_static = Int[]
+    interpolations_static = ScalarInterpolation[]
+
+    for group in IterTools.groupby(row -> row.subgrid_id, static)
         subgrid_id = first(getproperty.(group, :subgrid_id))
-        node_id = NodeID(NodeType.Basin, first(getproperty.(group, :node_id)), node_table)
+        node_id = first(getproperty.(group, :node_id))
         basin_level = getproperty.(group, :basin_level)
         subgrid_level = getproperty.(group, :subgrid_level)
 
         is_valid =
             valid_subgrid(subgrid_id, node_id, node_to_basin, basin_level, subgrid_level)
+        !is_valid && error("Invalid Basin / subgrid table.")
+
+        # Ensure it doesn't extrapolate before the first value.
+        pushfirst!(subgrid_level, first(subgrid_level))
+        pushfirst!(basin_level, nextfloat(-Inf))
+        hh_itp = LinearInterpolation(
+            subgrid_level,
+            basin_level;
+            extrapolate = true,
+            cache_parameters = true,
+        )
+        push!(subgrid_id_static, subgrid_id)
+        push!(basin_index_static, node_to_basin[node_id])
+        push!(interpolations_static, hh_itp)
+    end
 
-        if is_valid
-            # Ensure it doesn't extrapolate before the first value.
-            pushfirst!(subgrid_level, first(subgrid_level))
-            pushfirst!(basin_level, nextfloat(-Inf))
-            new_interp = LinearInterpolation(
-                subgrid_level,
-                basin_level;
-                extrapolate = true,
-                cache_parameters = true,
-            )
-            push!(subgrid_ids, subgrid_id)
-            push!(basin_index, node_to_basin[node_id])
-            push!(interpolations, new_interp)
-        else
-            has_error = true
+    subgrid_id_time = Int32[]
+    basin_index_time = Int[]
+    interpolations_time = ScalarInterpolation[]
+    current_interpolation_index = IndexLookup[]
+
+    # Push the first subgrid_id and basin_index
+    if length(time) > 0
+        push!(subgrid_id_time, first(time.subgrid_id))
+        push!(basin_index_time, node_to_basin[first(time.node_id)])
+    end
+
+    # Initialize index_lookup contents
+    lookup_time = Float64[]
+    lookup_index = Int[]
+
+    interpolation_index = 0
+    for group in IterTools.groupby(row -> (row.subgrid_id, row.time), time)
```

Comment:
On first review I didn't notice this groups on subgrid_id *and* time, which is why I was confused in the rest of the codebase. Still, this part is complex, and requires specific starting states (not just empty vectors). I'd appreciate a small text (or diagram?) on what/how you're building exactly.

Reply:
Yeah this is not simple stuff. I added more notes in dcf6182ecb93e324f43609f956bc4e3189eb92aa, and directly realized the docs on sorting was wrong (the code was correct).